doc/opdatasigverify.md

```diff
@@ -0,0 +1,118 @@
+
+Otherwise, the top 2 items are popped off the stack (leaving *data* on the top of the stack), and the opcode succeeds.
+
+## Examples
+
+* Simplest DATASIGVERIFY:
+  * output script:  `pubkeyhash OP_DATASIGVERIFY`
+  * spend script: `data sig`
+
+* Realistic example: Verify that `data` is signed by `dpubkeyhash` (signature `daddrsig`) and equals `x`.  Also do normal p2pkh transaction verification on `pubkeyhash` (with signature `txSig` and public key `pubkey`):
+  * output script: `dpubkeyhash OP_DATASIGVERIFY x OP_EQUALVERIFY OP_DUP OP_HASH160 pubkeyhash OP_EQUALVERIFY OP_CHECKSIG
+  * spend script: `txSig pubkey data daddrsig`
+
+## Discussion
+
+### Why is the non-verify version (e.g. OP_CHECKDATASIG) unnecessary?
+
+An instruction like OP_CHECKDATASIG would validate the signature on data and push a true/false result on the stack, analguous to OP_CHECKSIG vs OP_CHECKSIGVERIFY.
+
+A Bitcoin Cash script is not exactly like a normal program.  Its fundamental and only purpose is to encumber spendability -- your input script must meet the output script's requirements in order to spend the coins.  Therefore, any program sequence that does not narrow the group of possible spenders is equivalent to a simple choice.  In the OP_CHECKDATASIG case, any potential spender can create a spend script that fails OP_CHECKDATASIG by pushing random bytes rather than a valid signature, so the "false" result is equivalent to (but less efficient than) an if statement on a pushed constant.
+
+Formally, any script with encumberances x, y, z and solutions x', y', and z', of the form:
+
+`pubkeyhash OP_CHECKDATASIG OP_IF x OP_ELSE y OP_ENDIF z`
+
+if-branch spend: `z' x' data sig`
+
+else-branch spend: `z' y' data badsig`
+
+Can be replaced by the script:
+
+`OP_IF pubkeyhash OP_DATASIGVERIFY x OP_ELSE y OP_ENDIF z`
+
+if-branch spend:  `z' x' data sig 1`
+
+else-branch spend: `z' y' 0`
+
+
+## Reference Implementation
+
+Please refer to [this github branch](https://github.com/gandrewstone/BitcoinUnlimited/tree/op_datasigverify) for a complete implementation.  Implementation is in src/test/interpreter.cpp/h and unit tests are located at src/test/script_tests.cpp.
+
+The opcode implementation is short enough to include here:
+```c++
+// This code sits inside the interpreter's opcode processing case statement
+case OP_DATASIGVERIFY:
+{
+    if (stack.size() < 3)
+        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+    valtype &data = stacktop(-3);
+    valtype &vchSigAndType = stacktop(-2);
+    valtype &vchAddr = stacktop(-1);
+
+    if (vchAddr.size() != 20)
+        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+    if (vchSigAndType[0] == DATASIG_COMPACT_ECDSA)
```

Comment:
This should be at the end to be consistent with the other verification ops.  I wish that they had all been at the beginning for ease of use with OP_SPLIT, but consistency prevents more bugs.

Reply:
ok, I switched it to the end.